### PR TITLE
refactor[encoder]: Reuse single buffer

### PR DESCRIPTION
**Pull request by Yun-Kim**
### Current MsgpackEncoder: List Buffered Encoder

Currently, the encoder creates a new instance of `Packer()` to encode each trace, and then adds to a list buffer. In order to flush those encoded traces to the agent, this list buffer must then be joined into a single msgpack array at the end, which is a quite costly operation.

### Proposed MsgpackEncoder: Single Buffer Encoder

The proposed change is to keep and reuse the same msgpack buffer per each `MsgpackEncoder` instance, and to remove the list buffer (and therefore the costly `join()` operation) in favour of appending to the msgpack buffer for each trace that is encoded.  The array size prefix is left as an empty offset until it is modified right before flushing the encoded traces.

### Additional: MsgpackEncoderBase Abstraction

This change also includes abstracting out a large portion of the current `MsgpackEncoder` into the `MsgpackEncoderBase` class, which provides the encoder interface for `put()` and `encode()`, (note: `_pack_trace()` is provided as well by this base class, but is dependent on the implementation of `pack_span()`). The `MsgpackEncoder` class which inherits off of  `MsgpackEncoderBase` handles the implementation details behind `pack_span()`. The reason for this change was to provide a base interface for all of our msgpack encoders to follow, thus making any new encoders (i.e. `MsgpackEncoderV05`) easy to implement and compare with the current encoder.

### CODE REVIEW

**Review by brettlangdon — COMMENTED** [review]
it might be worth pulling the change to move `pack_number` and `pack_text` to module level in it's own PR, might help clean up this PR a bit.

File: ddtrace/internal/_encoding.pyx [hunk 1]
```diff
@@ -46,120 +46,225 @@ cdef inline int PyBytesLike_Check(object o):
...
-        packer = Packer()
-        astream.write(packer.pack_trace(trace))
-        astream.write(packer.pack_traces(traces))
+    def __cinit__(self, max_size, max_item_size):
+        self.max_size = max_size
+        self.max_item_size = max_item_size
+        self._lock = threading.Lock()
+
+    def put(self, item): ...
+    def encode(self): ...
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/internal/_encoding.pyx:134]: I think we've had this pattern for a while and had no problems, but I can just change these to raise `NotImplementedError()` since these are abstract methods anyway.

[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/internal/_encoding.pyx:134]: /shrug

File: ddtrace/internal/_encoding.pyx [hunk 2]
```diff
@@ -46,120 +46,225 @@ cdef inline int PyBytesLike_Check(object o):
...
+    def size(self):
+        """Return the size in bytes of the encoder buffer."""
+        return self.pk.length + array_prefix_size(self._count) - 5
+
+    # ---- Abstract methods ----
+
+    cpdef flush(self):
+        raise NotImplementedError()
+
+    cpdef pack_span(self, object span, object dd_origin):
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/internal/_encoding.pyx]: Tried changing to `cdef` and ran tests no problem, so 🤷

File: ddtrace/internal/_encoding.pyx [hunk 3]
```diff
@@ -46,120 +46,225 @@ cdef inline int PyBytesLike_Check(object o):
...
-        elif PyFloat_Check(n):
-            return msgpack_pack_double(&self.pk, <double> n)
-
-        raise TypeError("Unhandled numeric type: %r" % type(n))
-
-    cdef inline int _pack_text(self, object text):
+    def __len__(self):  # TODO: Use a better name?
+        return self._count
+
+    cpdef _decode(self, data):  # TODO: Make module util/class method?
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/internal/_encoding.pyx]: This is actually shared by both the current encoder and v0.5. The reason is that we only care how we encode things for efficiency, but decoding is general. That's why I've put that TODO there as perhaps this makes more sense on module level? 🤷

File: ddtrace/internal/_encoding.pyx (hunk 2 above)
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/internal/_encoding.pyx]: I think this was needed for v0.5 but would have to double check. If `cdef` works here then we should use that and change it in the v0.5 PR if required.

File: tests/benchmarks/test_encoding.py [hunk 4]
```diff
@@ -3,7 +3,7 @@
 import pytest
 
 from ddtrace.ext.ci import CI_APP_TEST_ORIGIN
-from ddtrace.internal.encoding import MsgpackEncoder
+from ddtrace.internal.encoding import Encoder as MsgpackEncoder
```
Comment (P403n1x87) [inline tests/benchmarks/test_encoding.py:6]: Is this change required for this PR? If not I'd keep the explicit import of `MsgpackEncoder` to avoid breaking tests if the default value of the `Encoder` variable should ever change. We'll then sort this out in the v0.5 PR.

File: tests/utils.py [hunk 5]
```diff
@@ -14,10 +14,10 @@
 from ddtrace import Tracer
 from ddtrace.constants import SPAN_MEASURED_KEY
 from ddtrace.ext import http
-from ddtrace.internal._encoding import MsgpackEncoder
 from ddtrace.internal.compat import httplib
 from ddtrace.internal.compat import parse
 from ddtrace.internal.compat import to_unicode
+from ddtrace.internal.encoding import Encoder
```
Comment (P403n1x87) [inline tests/utils.py:20]: > Is this change required for this PR? If not I'd keep the explicit import of `MsgpackEncoder` to avoid breaking tests if the default value of the `Encoder` variable should ever change. We'll then sort this out in the v0.5 PR.

Same here

File: ddtrace/internal/_encoding.pyx [hunk 6]
```diff
@@ -46,120 +48,230 @@ cdef inline int PyBytesLike_Check(object o):
...
+            ret = msgpack_pack_raw(pk, L)
+            if ret == 0:
+                ret = msgpack_pack_raw_body(pk, <char *> text, L)
+        return ret
+
+    raise TypeError("Unhandled text type: %r" % type(text))
+
 
-    usage::
+cdef class BufferedEncoder(object):
```
Comment (P403n1x87) [inline ddtrace/internal/_encoding.pyx:125]:
```suggestion
cdef class BufferedEncoder(object):
    content_type: str = None
```
My bad. I've realised afterwards that the `content_type` attribute is also used by the AppSec implementations of `BufferedEncoder` so I should have left it here 🙁 .

File: ddtrace/internal/_encoding.pyx [hunk 7]
```diff
@@ -1,6 +1,6 @@
 from cpython cimport *
 from cpython.bytearray cimport PyByteArray_Check
-import struct
+from libc cimport stdint
 import threading
 
 from ..constants import ORIGIN_KEY
```
Comment (P403n1x87) [inline ddtrace/internal/_encoding.pyx:6]: one further small optimisation would be to "unbox" this Python string to get hold of the underlying `char *` buffer so that we can call `msgpack_pack_raw*` directly on it instead of going through the heavier `pack_text`.

File: ddtrace/internal/_encoding.pyx [hunk 8]
```diff
@@ -46,120 +48,230 @@ cdef inline int PyBytesLike_Check(object o):
...
+
+        L = len(trace)
+        if L > ITEM_LIMIT:
+            raise ValueError("list is too large")
+
+        ret = msgpack_pack_array(&self.pk, L)
+        if ret != 0: raise RuntimeError("Couldn't pack trace")
+
+        if L > 0 and trace[0].context is not None:
+            dd_origin = trace[0].context.dd_origin
```
Comment (P403n1x87) [inline ddtrace/internal/_encoding.pyx:216]: Same here. We should see if it's possible to unbox this once and use `msgpack_pack_raw*` instead of `pack_text`. My only concern is with `pack_span`, but given that here it is still `cdef`ined we should be good.

File: ddtrace/internal/_encoding.pyx (hunk 7 above)
Comment (Yun-Kim) [reply ddtrace/internal/_encoding.pyx:6]: I'm thinking we just remove this import entirely and replace the ``pack_text(&self.pk, ORIGIN_KEY)`` with:
```python
ret = pack_bytes(&self.pk, <char *> b"_dd.origin", 10)
```
If we're doing the same for `span_id`, `trace_id`, `parent_id`, `service`, `resource`, `name`, `error`, `start`, `duration`, `type`, `meta`, `metrics`, I don't see why we shouldn't also just hard code the bytestring for dd_origin as well. @brettlangdon @P403n1x87 What do you think?

Comment (brettlangdon) [reply ddtrace/internal/_encoding.pyx:6]: 👍🏻 this is a good change to make

File: ddtrace/internal/_encoding.pyx (hunk 8 above)
Comment (brettlangdon) [reply ddtrace/internal/_encoding.pyx:217]: you mean something like:

```suggestion
            dd_origin = <char*> trace[0].context.dd_origin
```

?

pass around the c string rather than pulling it from the PyUnicode every time?

Comment (P403n1x87) [reply ddtrace/internal/_encoding.pyx:217]: yes!

File: ddtrace/internal/_encoding.pyx (hunk 7 above)
Comment (P403n1x87) [reply ddtrace/internal/_encoding.pyx]: Agreed. Presumably `_dd.origin` is unlikely to change, but in case it does we wouldn't have to worry about updating the string here too if we used the import.

File: ddtrace/internal/_encoding.pyx [hunk 9]
```diff
@@ -213,10 +215,12 @@ cdef class MsgpackEncoderBase(BufferedEncoder):
...
-        if L > 0 and trace[0].context is not None:
-            dd_origin = trace[0].context.dd_origin
-        else:
-            dd_origin = None
+        if L > 0 and trace[0].context is not None and trace[0].context.dd_origin is not None:
+            IF PY_MAJOR_VERSION >= 3:
+                dd_origin_bytestring = trace[0].context.dd_origin.encode("UTF-8")
+                dd_origin = dd_origin_bytestring
+            ELSE:
+                dd_origin = trace[0].context.dd_origin
```
Comment (brettlangdon) [inline ddtrace/internal/_encoding.pyx:223]: I _think_ you can just do:

```
dd_origin = <char*> trace[0].context.dd_origin
```

and then Cython handles the rest. you can compile with `cythonize(..., annotate=True)` to see what it generates, iirc it will do the coercion for you.

Comment (brettlangdon) [reply ddtrace/internal/_encoding.pyx:223]: we checked, no you cannot. handling of `bytes` is not handled by casting alone.

@Yun-Kim can you add a comment here to explain that?

Comment (Yun-Kim) [reply ddtrace/internal/_encoding.pyx]: Yes! I am actually going to pull those dd_origin improvement changes into a separate PR.

File: ddtrace/internal/_encoding.pyx (hunk 8 above)
Comment (Yun-Kim) [reply ddtrace/internal/_encoding.pyx:219]: I'll pull out the dd_origin improvement changes into a separate PR so we can keep this one as small as possible :)

**Review by P403n1x87 — DISMISSED** [review]

**Comment by Yun-Kim** [comment]
From the benchmarks encoding test scenario, there were no significant differences between the current and proposed encoder implementations.

**Comment by brettlangdon** [comment]
> From the benchmarks encoding test scenario, there were no significant differences between the current and proposed encoder implementations.

Does that make these changes valuable, at least the one to use a single buffer? compared to the refactor work

**Comment by Yun-Kim** [comment]
> Does that make these changes valuable, at least the one to use a single buffer? compared to the refactor work

Update: I ran the encoder test scenario again (missed a configuration in the previous run), and got the following results:
```
+----------------------------------------------------------+------------------------+-----------------------+
| Benchmark                                                | 0.52.0 (refactor)      | 0.51.2 (master)       |
+==========================================================+========================+=======================+
| scenario:encoder|ntraces:1|nspans:1000                   | 525 us                 | 487 us: 1.08x faster  |
+----------------------------------------------------------+------------------------+-----------------------+
| scenario:encoder|ntraces:100|nspans:1000                 | 54.5 ms                | 53.1 ms: 1.03x faster |
+----------------------------------------------------------+------------------------+-----------------------+
| scenario:encoder|ntraces:1|nspans:1000|ntags:48|ltags:64 | 3.56 ms                | 3.75 ms: 1.05x slower |
+----------------------------------------------------------+------------------------+-----------------------+
| scenario:encoder|ntraces:1|nspans:1000|nmetrics:1        | 662 us                 | 740 us: 1.12x slower  |
+----------------------------------------------------------+------------------------+-----------------------+
| Geometric mean                                           | (ref)                  | 1.01x slower          |
+----------------------------------------------------------+------------------------+-----------------------+
```
From this, it looks like that any traces/spans with any tags or metrics (so in a realistic scenario) the new refactored encoder performs significantly faster than the current encoder.

File: ddtrace/internal/_encoding.pyx [hunk 10]
```diff
@@ -46,16 +49,24 @@ cdef inline int PyBytesLike_Check(object o):
     return PyBytes_Check(o) or PyByteArray_Check(o)
 
 
+cdef inline int array_prefix_size(int l):
+    if l < 16:
+        return 1
+    elif l < (2<<16):
+        return 3
+    return MSGPACK_ARRAY_LENGTH_PREFIX_SIZE
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/internal/_encoding.pyx:57]: I can't think of a reason not to just use 5 other than to avoid unused space in the buffer, @P403n1x87 do you have any thoughts?

File: ddtrace/internal/_encoding.pyx [hunk 11]
```diff
@@ -112,38 +123,158 @@ cdef inline int pack_text(msgpack_packer *pk, object text):
...
     cdef msgpack_packer pk
-    cdef const char *encoding
-    cdef const char *unicode_errors
+    cdef stdint.uint32_t _count
 
-    def __cinit__(self):
-        cdef int buf_size = 1024*1024
-        self.pk.buf = <char*> PyMem_Malloc(buf_size)
+    def __cinit__(self, max_size, max_item_size):
+        self.pk.buf = <char*> PyMem_Malloc(max_size)
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/internal/_encoding.pyx:154]: We discussed this off of Zoom.

`msgpack_pack_write` will already handle Remalloc when the buffer size needs to grow.

Instead of pre-allocating 8mb all at once which not everyone will need, we will revert this back to the 1mb initial allocation we were doing before.

We can defer discussions on increasing that initial allocation and whether we want to disable the ability to remalloc the buffer size in the future.

File: ddtrace/internal/_encoding.pyx [hunk 12]
```diff
@@ -112,38 +123,160 @@ cdef inline int pack_text(msgpack_packer *pk, object text):
...
     def __dealloc__(self):
         PyMem_Free(self.pk.buf)
         self.pk.buf = NULL
 
-    cdef inline object _flush_buffer(self):
-        buf = PyBytes_FromStringAndSize(self.pk.buf, self.pk.length)
-        # Reset the buffer.
-        self.pk.length = 0
-        return buf
+    def __len__(self):  # TODO: Use a better name?
```
Comment (brettlangdon) [inline ddtrace/internal/_encoding.pyx:169]: what is wrong with this name?

**Review by brettlangdon — APPROVED** [review]

File: ddtrace/internal/_encoding.pyx (hunk 10 above)
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/internal/_encoding.pyx:57]: If we used 5 every time we would have to manually implement writing the size prefix according to the msgpack specs. If we use the C API for that we need this variable size. I opted for the latter to avoid leaking any msgpack implementation details into our code.

File: ddtrace/internal/_encoding.pyx (hunk 11 above)
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/internal/_encoding.pyx]: I had added a comment in the v0.5 PR about this point

https://github.com/DataDog/dd-trace-py/blob/2970a5fa0e6d181d5c4733abe883065db77426b8/ddtrace/internal/_encoding.pyx#L399-L403

my main concern there is in the case of a very large trace being encoded, which would automatically reallocate the buffer to the larger size. Of course this could be prevented by checking the current buffer utilisation at each write operation, but that adds overheads, especially if the case of large traces is rare.

File: ddtrace/internal/_encoding.pyx (hunk 12 above)
Comment (P403n1x87) [reply ddtrace/internal/_encoding.pyx:169]: The buffered encoder keeps track of the number of traces that have been encoded as well as the current buffer utilisation. Here `__len__` is used for the number of traces but could be easily confused for the buffer utilisation, hence why I wasn't particularly happy with this choice.

**Review by P403n1x87 — APPROVED** [review]
